File: apps/athenaeum/packages/domain/src/search-rpc.ts
import * as Schema from "effect/Schema"
import { EntityId } from "./node.js"

// Plan §"Full-text search" (resolved by the Decisions stage: "Search is a server RPC over
// WorkspaceDurableObject's DO SQLite, uniform for web and native... Implement graph_text_search as
// a real FTS5 virtual table... a searchNodes RPC method"). Wire schemas for that RPC — kept in
// their own file rather than folded into graph-rpc.ts because `graph_text_search` is
// deliberately *not* reachable through the general `RunViewInput`/`ViewSpec` compiler (see
// `read-model.ts`'s `compileRunView` doc comment for why: FTS5 `MATCH` query syntax doesn't fit
// the `eq`/`in`/`hasTag` predicate grammar `ViewPredicate` models), so this is a distinct,
// narrower RPC surface, not a `GraphViewName` case of the general one.

export class SearchNodesInput extends Schema.Class<SearchNodesInput>("SearchNodesInput")({
  workspaceId: EntityId,
  query: Schema.String.pipe(Schema.minLength(1)),
  // Optional so callers don't have to think about a limit for the common case; the backend
  // still clamps whatever's supplied (or defaulted) to its own hard cap — see `read-model.ts`'s
  // `MAX_SEARCH_RESULTS`.
  limit: Schema.optional(Schema.Number.pipe(Schema.int(), Schema.positive()))
}) {}

/** One search hit: the matching node's id/title plus a short excerpt of the matched page body
 *  (empty string if the node has no page, i.e. the match was on `title` alone). */
export class SearchResultEntry extends Schema.Class<SearchResultEntry>("SearchResultEntry")({
  nodeId: EntityId,
  title: Schema.String,
  snippet: Schema.String
}) {}

export class SearchNodesOutput extends Schema.Class<SearchNodesOutput>("SearchNodesOutput")({
  results: Schema.Array(SearchResultEntry)
}) {}

File: apps/athenaeum/native/AthenaeumDomain/Sources/AthenaeumDomain/SearchRPC.swift
import Foundation

/// Mirrors `packages/domain/src/search-rpc.ts`'s `SearchNodesInput` — `limit` optional, backend
/// clamps to its own hard cap regardless.
public struct SearchNodesInput: Codable, Hashable, Sendable {
    public let workspaceId: EntityId
    public let query: String
    public let limit: Int?

    public init(workspaceId: EntityId, query: String, limit: Int? = nil) {
        self.workspaceId = workspaceId
        self.query = query
        self.limit = limit
    }
}

/// Mirrors `search-rpc.ts`'s `SearchResultEntry` — one search hit: matching node's id/title plus a
/// short excerpt (empty string if the match was on `title` alone, no page body).
public struct SearchResultEntry: Codable, Hashable, Sendable {
    public let nodeId: EntityId
    public let title: String
    public let snippet: String

    public init(nodeId: EntityId, title: String, snippet: String) {
        self.nodeId = nodeId
        self.title = title
        self.snippet = snippet
    }
}

public struct SearchNodesOutput: Codable, Hashable, Sendable {
    public let results: [SearchResultEntry]
    public init(results: [SearchResultEntry]) { self.results = results }
}
